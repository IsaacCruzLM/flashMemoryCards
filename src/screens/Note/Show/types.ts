import {DefaultFormProps} from '../../../components/Form/types';
import {NoteModelType} from '../../../databases/models/noteModel';
import {CategoryModelType} from '../../../databases/models/categoryModel';

export interface ShowProps extends DefaultFormProps {
  route: object;
  note: NoteModelType;
  categories: CategoryModelType[];
}
