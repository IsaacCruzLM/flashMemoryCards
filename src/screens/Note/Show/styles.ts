import {StyleSheet} from 'react-native';

const styles = StyleSheet.create({
  container: {
    flex: 1,
    width: '100%',
    flexDirection: 'column',
    justifyContent: 'space-between',
  },
});

export default styles;
