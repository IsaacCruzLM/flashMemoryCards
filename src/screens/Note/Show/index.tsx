import React, {useContext, useState} from 'react';
import {View} from 'react-native';
import withObservables from '@nozbe/with-observables';
import {withDatabase} from '@nozbe/watermelondb/DatabaseProvider';
import {compose} from 'recompose';
import get from 'lodash/get';
import cloneDeep from 'lodash/cloneDeep';

import DefaultContainerView from '../../../components/DefaultContainerView';
import TextEditor from '../../../components/TextEditor';
import Button from '../../../components/Button';
import Dialog from '../../../components/Dialog';
import TextInput from '../../../components/TextInput';
import Select from '../../../components/Select';
import SelectMultiple from '../../../components/SelectMultiple';

import AppContext from '../../../context/appContext';
import WmdbUtils from '../../../databases/utils';
import NavigationService from '../../../navigation/NavigationService';
import {translateOptions} from '../Create';

import styles from './styles';
import {ShowProps} from './types';
import {CategoryModelType} from '../../../databases/models/categoryModel';

const Show: React.FunctionComponent<ShowProps | any> = ({
  route,
  note,
  categories,
  subjects,
}) => {
  const {globalState} = useContext(AppContext);
  const [newContent, setNewContent] = useState(note.content);
  const [newInfo, setNewInfo] = useState({
    name: get(note, 'name', ''),
  });

  const submitAction = async () => {
    const noteId = get(route, 'params.noteId', '');
    const category = categories.find(
      ({id}: CategoryModelType) => id === note.category.id,
    );
    await WmdbUtils.updateItemInWMDB('notes', {content: newContent}, noteId);
    NavigationService.navigate('Notes', {
      categoryName: get(category, 'name', ''),
      categoryId: get(category, 'id', ''),
    });
  };

  const handleInfoChange = (key: 'name', value: string) => {
    const newInfoClone = cloneDeep(newInfo);
    newInfoClone[key] = value;
    setNewInfo(newInfoClone);
  };

  return (
    <DefaultContainerView>
      <View style={styles.container}>
        <View>
          <TextEditor
            initialContentHTML={note.content}
            onChange={content => setNewContent(content)}
            placeHolder="Escreva sua anotação aqui!"
          />
        </View>
        {globalState.keyboardIsVisible ? null : (
          <View>
            <Button label={'Salvar Anotação'} onPress={submitAction} />
          </View>
        )}
      </View>
      <Dialog
        actions={[
          {
            label: 'Atuzalizar',
            buttonMode: 'contained',
            buttonAction: () => {},
          },
          {
            label: 'Fechar',
            buttonMode: 'outlined',
            buttonAction: () => {},
          },
        ]}
        isVisible={true}
        hideDialog={() => {}}
        title={'AA'}>
        <View>
          <TextInput
            label={'Nome da anotação'}
            setText={value => handleInfoChange('name', value)}
            placeholder={'Nome da anotação'}
            value={newInfo.name}
          />
          <Select
            options={translateOptions(categories)}
            onChange={value => {}}
            modalTitle="Selecione uma categoria"
            inputLabel="Selecionar Categoria"
            inputPlaceHolder="Selecionar Categoria"
            defaultValue={''}
          />
          <SelectMultiple
            options={translateOptions(subjects)}
            onChange={value => {}}
            modalTitle="Selecione vários assuntos"
            inputPlaceHolder="Selecionar Assuntos"
          />
        </View>
      </Dialog>
    </DefaultContainerView>
  );
};

export default compose(
  withDatabase,
  withObservables([], ({route, database}: any) => {
    const noteId = get(route, 'params.noteId', '');
    return {
      note: database.get('notes').findAndObserve(noteId),
      categories: database.get('categories').query(),
      subjects: database.get('subjects').query(),
    };
  }),
)(Show);
