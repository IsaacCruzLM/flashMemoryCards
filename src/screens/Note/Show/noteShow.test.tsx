import React from 'react';
import renderer from 'react-test-renderer';
import Show from './index';

describe('Note Show Screen', () => {
  let tree: renderer.ReactTestRenderer;
  beforeEach(() => {
    jest.clearAllMocks();

    tree = renderer.create(<Show />);
  });

  afterEach(() => {
    jest.clearAllMocks();
    jest.useRealTimers();
  });

  it('should render correctly', () => {
    expect(tree.toJSON()).toBeTruthy();
    expect(tree.toJSON()).toMatchSnapshot();
  });

  it('render action button correctly', () => {
    const root = tree.root;

    const buttonToCancel = root.findByProps({
      testID: 'ButtonTestID Salvar Anotação',
    });

    expect(buttonToCancel).toBeTruthy();
  });
});
