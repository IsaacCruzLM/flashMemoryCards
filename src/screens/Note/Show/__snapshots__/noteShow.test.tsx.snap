// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`Note Show Screen should render correctly 1`] = `
<View
  style={
    Array [
      Object {
        "alignItems": "center",
        "flex": 1,
        "justifyContent": "space-between",
        "paddingHorizontal": 20,
        "paddingVertical": 24,
        "width": "100%",
      },
      undefined,
    ]
  }
>
  <View
    style={
      Object {
        "flex": 1,
        "flexDirection": "column",
        "justifyContent": "space-between",
        "width": "100%",
      }
    }
  >
    <View>
      <View>
        <Text>
          Erro
        </Text>
      </View>
    </View>
    <View>
      <View
        accessibilityRole="button"
        accessibilityState={
          Object {
            "disabled": false,
          }
        }
        accessible={true}
        collapsable={false}
        focusable={true}
        onBlur={[Function]}
        onClick={[Function]}
        onFocus={[Function]}
        onResponderGrant={[Function]}
        onResponderMove={[Function]}
        onResponderRelease={[Function]}
        onResponderTerminate={[Function]}
        onResponderTerminationRequest={[Function]}
        onStartShouldSetResponder={[Function]}
        style={
          Array [
            Object {
              "overflow": "hidden",
            },
            false,
            Object {
              "borderRadius": 20,
            },
          ]
        }
        testID="ButtonTestID Salvar Anotação"
      >
        <View
          style={
            Array [
              Object {
                "alignItems": "center",
                "flexDirection": "row",
                "justifyContent": "center",
              },
              undefined,
            ]
          }
        >
          <Text
            numberOfLines={1}
            selectable={false}
            style={
              Array [
                Object {
                  "color": "rgba(28, 27, 31, 1)",
                  "fontFamily": "System",
                  "fontSize": 14,
                  "fontWeight": "500",
                  "letterSpacing": 0.1,
                  "lineHeight": 20,
                },
                Object {
                  "textAlign": "left",
                },
                Object {
                  "writingDirection": "ltr",
                },
                Array [
                  Object {
                    "marginHorizontal": 16,
                    "marginVertical": 9,
                    "textAlign": "center",
                  },
                  false,
                  Object {
                    "marginHorizontal": 24,
                    "marginVertical": 10,
                  },
                  undefined,
                  false,
                  Object {
                    "color": "#FFFFFF",
                    "fontFamily": "System",
                    "fontSize": 14,
                    "fontWeight": "500",
                    "letterSpacing": 0.1,
                    "lineHeight": 20,
                  },
                  undefined,
                ],
              ]
            }
          >
            Salvar Anotação
          </Text>
        </View>
      </View>
    </View>
  </View>
</View>
`;
